### PR TITLE
Optimize JSON-Proto Conversion, Improve Memory Allocation, and Add Unit Tests

### PR DESCRIPTION
**Pull request by redouan-rhazouani**
- Memory Optimization: Improved allocation by eliminating raw loops.
- Code Readability: Enhanced code readability by reducing nesting and restructuring logic for clarity.
- Unit Tests: Added unit tests to assert correctness and reliability of the JSON-Proto conversion.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by timvisee — COMMENTED** [review]
Thank you very much! :pray: 

Out of curiosity, did you patch this to fix a measurable performance problem, or did you intend this to be just a nice to have?

Either way, thanks for the changes. I like the new loops with proper pre-allocation, and thanks for the tests.

Would you mind to run clippy to make the CI tests green. It may just be this one line below:

File: lib/api/src/conversions/json.rs [hunk 1]
```diff
@@ -58,51 +47,193 @@ pub fn json_path_from_proto(a: &str) -> Result<JsonPath, Status> {
...
-    let mut map: serde_json::Map<String, serde_json::Value> = serde_json::Map::new();
-    for (k, v) in proto {
-        map.insert(k, proto_to_json(v)?);
-    }
-    Ok(map.into())
+    proto
+        .into_iter()
+        .map(|(k, v)| proto_to_json(v).map(|json| (k, json)))
+        .collect::<Result<serde_json::Map<_, _>, _>>()
+        .map(|m| segment::types::Payload(m))
```
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply lib/api/src/conversions/json.rs]: Thanks, Tim for the feedback! 😊 I'm passionate about vector DBs and Rust, so I try to contribute to OSS whenever I can. Glad you liked the changes! 🚀

**Comment by redouan-rhazouani** [comment]
> Thank you very much! 🙏
> 
> Out of curiosity, did you patch this to fix a measurable performance problem, or did you intend this to be just a nice to have?
> 
> Either way, thanks for the changes. I like the new loops with proper pre-allocation, and thanks for the tests.
> 
> Would you mind to run clippy to make the CI tests green. It may just be this one line below:

Thanks, Tim for the feedback! 😊 I'm passionate about vector DBs and Rust, so I try to contribute to OSS whenever I can. Glad you liked the changes! 🚀

**Review by timvisee — APPROVED** [review]
Sweet! Changes like this are always welcome.

Approved. I expect CI to go through now. Thanks again!

**Comment by timvisee** [comment]
> I expect CI to go through now.

Whoops, there seem to be some more.

Please run `cargo clippy --workspace --all-targets -- -D warnings` locally to ensure all lint warnings are handled.

**Comment by redouan-rhazouani** [comment]
> > I expect CI to go through now.
> 
> Whoops, there seem to be some more.
> 
> Please run `cargo clippy --workspace --all-targets -- -D warnings` locally to ensure all lint warnings are handled.

Apologies for the inconvenience - this is my first contribution. I've resolved the issues, and it should work now. Thanks for pointing it out!

**Pull request merged**